lib/shoryuken/logging.rb

```diff
@@ -16,10 +16,19 @@ def context
     end
 
     def self.with_context(msg)
-      Thread.current[:shoryuken_context] = msg
+      if logger.respond_to?(:add_context)
+        logger.add_context(msg)
```

Comment:
👋 

What's the logger impl that supports `add_context`?

Reply:
We have an inherited `ActiveSupport::TaggedLogging` logger that we will add the custom method to.

Reply:
I'm wondering if that's specific for your custom method. Therefore it may not be generic enough to have it in Shoryuken.  WDYT?

If that's the case (too specific), you can extend the current logger, adding your specific code, then change the Shoryuken default logger. It is an option [here](https://github.com/phstc/shoryuken/blob/15923f8b02793c412dc430e16f1f4742d9e1f471/lib/shoryuken/options.rb#L89).

Reply:
we set a logger in our initializer, `Shoryuken::Logging.logger = Rails.logger` - this logger adds a lot of observability goodies for us and I imagine other companies who have multiple apps running ruby do the same.  

I think either shoryuken should establish an interface (like this PR suggests) for this added context or we remove it in favor of in-lining the context somehow. 

As it stands, we didn't even know this was available and while we'd prefer not to reach into `Thread.current`, I don't see a good way to do this without an interface.  I'm open to other suggestions if you have 'em

Reply:
Hi @allcentury 

How about nesting `with_context`? so we don't need to introduce a new interface?

```ruby
Shoryuken::Logging.with_context(msg_a) do 
  # ...
  Shoryuken::Logging.with_context(msg_b) do 
    # ...
  end
end
```

```ruby
def self.with_context(msg)
  Thread.current[:shoryuken_context] ||= []
  Thread.current[:shoryuken_context] << msg

  yield
ensure
  Thread.current[:shoryuken_context].pop
end
```

I have mixed feelings with `add_context`. I'm wondering if this is for error tracking if so, I believe that maybe error trackers could be more appropriated. [Honeybadger](https://github.com/honeybadger-io/honeybadger-ruby/blob/master/lib/honeybadger/plugins/shoryuken.rb) and [Rollbar](https://docs.rollbar.com/docs/shoryuken-integration) have built-in integrations with Shoryuken and both have methods for adding contexts. 

Just so I can get more context on the use-case, how are you using this logging?

Reply:
I understand your mixed feelings, is there a way to remove `Shoryuken::Logging.with_context` so there is no special ❄️ for `Shoryuken::Logging` vs using a different logger?

The use case is we don't want to miss out on valuable log lines because of which logger Shoryken is using.

Reply:
There's only one reference in the code for [`with_context`](https://github.com/phstc/shoryuken/blob/bc7938942270a1171d1597064698d88c3f36a4ec/lib/shoryuken/processor.rb#L19).

We have a few options:

1. In your workers add a `rescue` and use your own logging.

```ruby
def perform
rescue => e
  # your own logging
  raise
end
```

2. Create a middleware to ensure all your worker failures will be logged in the same way

```ruby
def call(_worker, queue, _sqs_msg, _body)
  custom_with_context(...) do
    yield
  end
rescue
  # moar logging
  raise
end
```

3. Use [my example](https://github.com/phstc/shoryuken/pull/597#discussion_r407112579) for nesting `with_context`. Wouldn't this work?

4. Make `Logging` a Shoryuken `Option`.

Reply:
1.  That could work but there are obvious tradeoffs there
2.  This is what we're doing today
3.  I don't grok the 3rd option, the nesting doesn't really change that the log formatter has this special method called `context` that knows to reach into a thread local variable https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/logging.rb#L13 .  I'm trying to come up with a way to make that generic so any logger could store or pass that valuable context information as needed.  
4. Not sure what you're thinking here

Reply:
3. 

With my change suggestion, this 👇 would possible.

```ruby
Shoryuken::Logging.with_context(...) do # internal from the `processor.rb`

   # Somewhere in your  code
   Shoryuken::Logging.with_context(...) do
      # this now will have the context added by the `process.rb` + the added context above
   end
end
```
Acting similarly to your `add_context`.

```ruby
Shoryuken::Logging.with_context(...) do # internal from the `processor.rb`

   # Somewhere in your  code
   Shoryuken::Logging.add_context(...)
   # this now will have the context added by the `process.rb` + the added context above
end
```

4. 

Another idea is to make `Shorykuen::Logging` customizable. So you can add your implementation that calls `add_context` or any other specifics to your use-case. 

You could:

<details>
  <summary>Sample code</summary>

```ruby
module Shoryuken
  module Logging
    def self.with_context(*args)
      Shoryuken.logging_handler.with_context(*args)
    end

    def self.initialize_logger(*args)
      Shoryuken.logging_handler.initialize_logger(*args)
    end

    def self.logger
      Shoryuken.logging_handler.logger
    end

    def self.logger=(log)
      Shoryuken.logging_handler.logger = log
    end
  end
end

module Shoryuken
  module DefaultLoggingHandler
    class Pretty < Logger::Formatter
      # Provide a call() method that returns the formatted message.
      def call(severity, time, _program_name, message)
        "#{time.utc.iso8601} #{Process.pid} TID-#{Thread.current.object_id.to_s(36)}#{context} #{severity}: #{message}\n"
      end

      def context
        c = Thread.current[:shoryuken_context]
        c ? " #{c}" : ''
      end
    end

    def self.with_context(msg)
      Thread.current[:shoryuken_context] = msg
      yield
    ensure
      Thread.current[:shoryuken_context] = nil
    end

    def self.initialize_logger(log_target = STDOUT)
      @logger = Logger.new(log_target)
      @logger.level = Logger::INFO
      @logger.formatter = Pretty.new
      @logger
    end

    def self.logger
      @logger || initialize_logger
    end

    def self.logger=(log)
      @logger = (log ? log : Logger.new('/dev/null'))
    end
  end
end

module Shoryuken
  class Options
    def logging_handler
      @logging_handler || DefaultLoggingHandler
    end
  end
end
```
</details>

☝️ with that, you could `Shoryuken.logging_handler = YourLoggingHandlerWithAddContext`.